include/deal.II/matrix_free/fe_remote_evaluation.h

```diff
@@ -0,0 +1,1120 @@
+              AssertIndexRange(j, dst.size());
+              AssertIndexRange(c, src.size());
+
+              copy_data(dst[j], v, src[c]);
+            }
+      }
+  }
+
+  /**
+   * Copy data between different data layouts.
+   */
+  template <typename T1, std::size_t n_lanes>
+  void
+  copy_data(VectorizedArray<T1, n_lanes> &dst,
+            const unsigned int            v,
+            const T1                     &src) const
+  {
+    AssertIndexRange(v, n_lanes);
+
+    dst[v] = src;
+  }
+
+  /**
+   * Copy data between different data layouts.
+   */
+  template <typename T1, int rank_, std::size_t n_lanes, int dim_>
+  void
+  copy_data(Tensor<rank_, dim_, VectorizedArray<T1, n_lanes>> &dst,
+            const unsigned int                                 v,
+            const Tensor<rank_, dim_, T1>                     &src) const
+  {
+    AssertIndexRange(v, n_lanes);
+
+    if constexpr (rank_ == 1)
+      {
+        for (unsigned int i = 0; i < dim_; ++i)
+          dst[i][v] = src[i];
+      }
+    else
+      {
+        for (unsigned int i = 0; i < rank_; ++i)
+          for (unsigned int j = 0; j < dim_; ++j)
+            dst[i][j][v] = src[i][j];
+      }
+  }
+
+  /**
+   * Copy data between different data layouts.
+   */
+  template <typename T1,
+            int         rank_,
+            std::size_t n_lanes,
+            int         n_components_,
+            int         dim_>
+  void
+  copy_data(
+    Tensor<rank_,
+           n_components_,
+           Tensor<rank_, dim_, VectorizedArray<T1, n_lanes>>>   &dst,
+    const unsigned int                                           v,
+    const Tensor<rank_, n_components_, Tensor<rank_, dim_, T1>> &src) const
+  {
+    if constexpr (rank_ == 1)
+      {
+        for (unsigned int i = 0; i < n_components_; ++i)
+          copy_data(dst[i], v, src[i]);
+      }
+    else
+      {
+        for (unsigned int i = 0; i < rank_; ++i)
+          for (unsigned int j = 0; j < n_components_; ++j)
+            dst[i][j][v] = src[i][j];
+      }
+  }
+
+  /**
+   * CRS like data structure that describes the data positions at given
+   * indices.
+   */
+  FERemoteEvaluationDataViewType view;
+  /**
+   * RemotePointEvaluation objects and indices to points used in
+   * RemotePointEvaluation.
+   */
+  typename FEETT::CommunicationObjectType communication_objects;
+};
+
+/**
+ * Class to access data in matrix-free loops for non-matching discretizations.
+ * Interfaces are named with FEEvaluation, FEFaceEvaluation or FEPointEvaluation
+ * in mind. The main difference is, that `gather_evaluate()` updates and caches
+ * all values at once. Therefore, it has to be called only once before a
+ * matrix-free loop.
+ *
+ * FERemoteEvaluation is thought to be used with another @p FEEvaluationType
+ * (FEEvaluation, FEFaceEvaluation, or FEPointEvaluation).
+ * FERemoteEvaluationCommunicator knows the type. However,
+ * FERemoteEvaluationCommunicator is independent of @p n_components.
```

Comment:
When is this useful?

Reply:
I restructured a lot of the code which resolves this.